packages/sanity/src/core/form/field/FieldActions.tsx

```diff
@@ -0,0 +1,106 @@
+import {Path, SchemaType} from '@sanity/types'
+import React, {memo, useCallback, useEffect, useRef, useState} from 'react'
+import {DocumentFieldAction, DocumentFieldActionNode} from '../../config'
+import {useUnique} from '../../util'
+
+/** @internal */
+export interface FieldActionsProps {
+  actions: DocumentFieldAction[]
+  documentId: string
+  documentType: string
+  onActions: (actions: DocumentFieldActionNode[]) => void
+  path: Path
+  schemaType: SchemaType
+}
+
+/** @internal */
+export const FieldActions = memo(function FieldActions(props: FieldActionsProps) {
+  const {actions, documentId, documentType, onActions, path, schemaType} = props
+
+  const len = actions.length
+  const lenRef = useRef(len)
+
+  const [fieldActions, setFieldActions] = useState<DocumentFieldActionNode[]>(() =>
+    Array.from(new Array(len))
+  )
+
+  const fieldActionsRef = useRef(fieldActions)
+
+  useEffect(() => {
+    fieldActionsRef.current = fieldActions
+  }, [fieldActions])
+
+  useEffect(() => {
+    if (lenRef.current !== len) {
```

Comment:
How will this work when actions are replaced but not added/removed in the api?

Reply:
That's handled by the `setFieldAction(index, node)` which is called whenever a changed/added action node is rendered.